From ecf638c03aa059696ff8e2718f3397449dcde2c4 Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Fri, 31 Aug 2018 23:23:12 +0530
Subject: [PATCH 0396/1851] drm: xlnx: sdi: Fix timing for 720x486i SD mode

commit e7637bd325177cac7a2ce15973b4b54ba60b098a from
https://github.com/Xilinx/linux-xlnx.git

Fixes the timing values and mode name for 720x486i SD mode.
The field 0 and 1 will have 243 and 244 active lines respectively in SD
mode.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_sdi_modes.h  | 6 +++---
 drivers/gpu/drm/xlnx/xlnx_sdi_timing.c | 7 ++++++-
 2 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/xlnx/xlnx_sdi_modes.h b/drivers/gpu/drm/xlnx/xlnx_sdi_modes.h
index 1f606ef6b71f..534f7d80f29c 100644
--- a/drivers/gpu/drm/xlnx/xlnx_sdi_modes.h
+++ b/drivers/gpu/drm/xlnx/xlnx_sdi_modes.h
@@ -37,9 +37,9 @@ struct xlnx_sdi_display_config {
 static const struct xlnx_sdi_display_config xlnx_sdi_modes[] = {
 	/* 0 - dummy, VICs start at 1 */
 	{ },
-	/* SD: 720x480i@60Hz */
-	{{ DRM_MODE("720x480i", DRM_MODE_TYPE_DRIVER, 13500, 720, 739,
-		   801, 858, 0, 240, 244, 247, 262, 0,
+	/* SD: 720x486i@60Hz */
+	{{ DRM_MODE("720x486i", DRM_MODE_TYPE_DRIVER, 13500, 720, 739,
+		   801, 858, 0, 243, 247, 250, 262, 0,
 		   DRM_MODE_FLAG_PHSYNC | DRM_MODE_FLAG_PVSYNC |
 		   DRM_MODE_FLAG_INTERLACE | DRM_MODE_FLAG_DBLCLK),
 		   .vrefresh = 60, }, {0x7, 0x6},
diff --git a/drivers/gpu/drm/xlnx/xlnx_sdi_timing.c b/drivers/gpu/drm/xlnx/xlnx_sdi_timing.c
index 9c9348c1ad9c..61ee98e87fdc 100644
--- a/drivers/gpu/drm/xlnx/xlnx_sdi_timing.c
+++ b/drivers/gpu/drm/xlnx/xlnx_sdi_timing.c
@@ -339,8 +339,13 @@ void xlnx_stc_sig(void __iomem *base, struct videomode *vm)
 	reg |= (vactive & XSTC_GA_ACTSIZE_MASK) << 16;
 	xlnx_stc_writel(base, XSTC_GASIZE, reg);
 
-	if (vm->flags & DISPLAY_FLAGS_INTERLACED)
+	if (vm->flags & DISPLAY_FLAGS_INTERLACED) {
+		if (vactive == 243)
+			reg = ((vactive + 1) & XSTC_GA_ACTSIZE_MASK) << 16;
+		else
+			reg = (vactive & XSTC_GA_ACTSIZE_MASK) << 16;
 		xlnx_stc_writel(base, XSTC_GASIZE_F1, reg);
+	}
 
 	reg = hsync_start & XSTC_GH1_SYNCSTART_MASK;
 	reg |= (hbackporch_start << XSTC_GH1_BPSTART_SHIFT) &
-- 
2.31.1

